iree/compiler/Dialect/LinalgExt/Transforms/ConvertToLoops.cpp

```diff
@@ -16,112 +16,93 @@
+                                      unsigned loopDepth,
+                                      SmallVectorImpl<Value> &ivs) {
+  Location loc = tilableOp.getLoc();
+  if (loopDepth == loopRanges.size()) {
+    return tilableOp.generateScalarImplementation(builder, loc, ivs);
   }
-};
-
-//===----------------------------------------------------------------------===//
-// ScatterOp
-//===----------------------------------------------------------------------===//
+  LogicalResult status = success(true);
+  builder.create<scf::ForOp>(
+      loc, loopRanges[loopDepth].offset, loopRanges[loopDepth].size,
+      loopRanges[loopDepth].stride, ValueRange{},
+      [&](OpBuilder &b, Location loc, Value iv, ValueRange args) {
+        ivs.push_back(iv);
+        status = lowerToLoopsImpl(b, tilableOp, loopRanges, loopDepth + 1, ivs);
+        b.create<scf::YieldOp>(loc);
+      });
+  return status;
+}
 
-struct ScatterConversion : public OpRewritePattern<linalg_ext::ScatterOp> {
-  using OpRewritePattern<linalg_ext::ScatterOp>::OpRewritePattern;
+/// Main entry point for lowering `TiledOpInterface` op to loops.
+static LogicalResult lowerToLoops(OpBuilder &builder,
+                                  TiledOpInterface tilableOp) {
+  SmallVector<Range> loopBounds = tilableOp.getLoopBounds(builder);
+  SmallVector<Value> ivs;
+  return lowerToLoopsImpl(builder, tilableOp, loopBounds, 0, ivs);
+}
 
-  LogicalResult matchAndRewrite(linalg_ext::ScatterOp op,
-                                PatternRewriter& rewriter) const final {
-    if (!op.hasBufferSemantics()) return failure();
-    auto updates = op.updates();
-    Location loc = op.getLoc();
-    Value zero = rewriter.create<ConstantIndexOp>(loc, 0);
-    Value one = rewriter.create<ConstantIndexOp>(loc, 1);
-    Value ub = rewriter.createOrFold<memref::DimOp>(loc, updates, zero);
-    bool fail = false;
-    rewriter.create<scf::ForOp>(
-        loc, zero, ub, one, ValueRange{},
-        [&](OpBuilder& b, Location loc, Value iv, ValueRange iters) {
-          if (failed(op.generateScalarImplementation(b, loc, iv))) {
-            fail = true;
-          }
-          b.create<scf::YieldOp>(loc);
-        });
-    if (fail) return failure();
+/// Pattern rewriter hook to lower a `TiledOpInterface` to loops.
+namespace {
+struct TiledOpInterfaceLowerToLoopsPattern : public RewritePattern {
+  TiledOpInterfaceLowerToLoopsPattern(MLIRContext *context,
+                                      PatternBenefit benefit = 1)
+      : RewritePattern(MatchAnyOpTypeTag(), benefit, context) {}
+
+  LogicalResult matchAndRewrite(Operation *op,
+                                PatternRewriter &rewriter) const override {
+    auto tilableOp = dyn_cast<TiledOpInterface>(op);
+    if (!tilableOp) {
+      return failure();
+    }
+    if (llvm::any_of(tilableOp->getOperands(), [&](Value v) {
+          return v.getType().isa<RankedTensorType>();
+        })) {
```

Comment:
nit: we can cast the op to LinalgExtOp and use `hasTensorSemantics` method. It could be scalar or RankedTensorType. In fft we do have the stage which is a scalar index type.

Reply:
Changed it to see if any of the results are shaped type. That should be a good enough proxy for `hasTensorSemantics`. Notice that the transformation here is not related to `LinalgExt` at all. So trying to keep that separate. We can revisit that if needed.